File: ch15/ex15-16.py
# 15.16 알 수 없는 인코딩의 C 문자열 다루기

s = retstr()
s
# 'Spicy Jalapeño\udcae'
print_chars(s)
# 53 70 69 63 79 20 4a 61 6c 61 70 65 c3 b1 6f ae



# 토론

raw = b'Spicy Jalape\xc3\xb1o\xae'
raw.decode('utf-8','ignore')
# 'Spicy Jalapeño'
raw.decode('utf-8','replace')
# 'Spicy Jalapeño?'


raw.decode('utf-8','surrogateescape')
# 'Spicy Jalapeño\udcae'


s = raw.decode('utf-8', 'surrogateescape')
print(s)
# Traceback (most recent call last):
#   File "<stdin>", line 1, in <module>
# UnicodeEncodeError: 'utf-8' codec can't encode character '\udcae' in position 14: surrogates not allowed


s
# 'Spicy Jalapeño\udcae'
s.encode('utf-8','surrogateescape')
# b'Spicy Jalape\xc3\xb1o\xae'

File: ch15/ex15-16.c
// 15.16 알 수 없는 인코딩의 C 문자열 다루기

/* 알 수 없는 문자열 데이터 (잘못된 UTF-8) */
const char *sdata = "Spicy Jalape\xc3\xc3\xb1o\xae";
int slen = 16;

/* 캐릭터 데이터 출력 */
void print_chars(char *s, int len) {
  int n = 0;
  while (n < len) {
    printf("%2x ", (unsigned char) s[n]);
    n++;
  }
  printf("\n");
}


/* C 문자열을 파이썬에 반환 */
static PyObject *py_retstr(PyObject *self, PyObject *args) {
  if (!PyArg_ParseTuple(args, "")) {
    return NULL;
  }
  return PyUnicode_Decode(sdata, slen, "utf-8", "surrogateescape");
}

/* print_char() 함수에 대한 래퍼 */
static PyObject *py_print_chars(PyObject *self, PyObject *args) {
  PyObject *obj, *bytes;
  char *s = 0;
  Py_ssize_t   len;

  if (!PyARG_ParseTuple(args, "U", &obj)) {
    return NULL;
  }

  if ((bytes = PyUnicode_AsEncodedString(obj,"utf-8","surrogateescape"))
      == NULL) {
    return NULL;
  }
  PyBytes_AsStringAndSize(bytes, &s, &len);
  print_chars(s, len);
  Py_DECREF(bytes);
  Py_RETURN_NONE;
}
